### ISSUE TITLE
[FR]: Add support for X3 MEGA Inverters

### ISSUE THREAD

**Issue by jiriknez** [wills106/homeassistant-solax-modbus#944]
### Description

[Max&Mega K-TL  Modbus Protocol V4.16.pdf](https://github.com/user-attachments/files/16043599/Max.Mega.K-TL.Modbus.Protocol.V4.16.pdf)

### Suggested Solution

Add new set of modbus register for X3 MEGA inverters.

**Comment by wills106** [comment]
I have added in initial support in [2024.07.0b1](https://github.com/wills106/homeassistant-solax-modbus/releases/tag/2024.07.0b1) but is your Mega a newer G2 version? As it might use different registers compared to that doc. That doc is quite old.

**Comment by jiriknez** [comment]
You're right, G2 version use different registers. I'm sending new document for Mega and Forth inverters. Please could you add these registers support? 

[Solax Power X3-MEGA G2_FORTH Modbus-RTU comms protocol (2023.5.16)-1.pdf](https://github.com/user-attachments/files/16138458/Solax.Power.X3-MEGA.G2_FORTH.Modbus-RTU.comms.protocol.2023.5.16.-1.pdf)

**Comment by wills106** [comment]
Just released [2024.07.0b5](https://github.com/wills106/homeassistant-solax-modbus/releases/tag/2024.07.0b5) does it return a serial number?

If it does then replace `MAXMEG_G2` with the first 4 or 6 digits of the serial number it returns.

https://github.com/wills106/homeassistant-solax-modbus/blob/2666d6320801bfeecc24bf2c28902113d4f5e525/custom_components/solax_modbus/plugin_solax.py#L7884-L7886

**Comment by jiriknez** [comment]
No, it doesn't read SN. 

custom_components/solax_modbus/plugin_solax.py:81
reading serial number from address 0x0 failed; other address may succeed
reading serial number from address 0x300 failed; other address may succeed

**Comment by jiriknez** [comment]
It seems that reading stops on address 0x1a10 and return blank value: 
Read SolaxMAX 0x1a10 serial number:     
unrecognized inverter type - serial number :

**Comment by jiriknez** [comment]
[home-assistant_2024-08-02T15-56-51.588Z.log](https://github.com/user-attachments/files/16472670/home-assistant_2024-08-02T15-56-51.588Z.log)

**Comment by wills106** [comment]
I have released [2024.08.1](https://github.com/wills106/homeassistant-solax-modbus/releases/tag/2024.08.1) where the Mega / Forth has it's own plugin. Can you try that and select solx_mega_forth from the dropdown.

**Comment by jiriknez** [comment]
[plugin_solax_mega_forth.zip](https://github.com/user-attachments/files/16487076/plugin_solax_mega_forth.zip)

Now it works, but I have problem with register with data type uint32 and int32. The values are too big. I also made same modifications in the plugin. Do you have idea why are values from 2bytes regiters read wrong?

**Comment by wills106** [comment]
Looks like I forgot to delete "Measure Power L1" I copied it for the unit_of_measurement and class
I wasn't sure if the docs were correct on the int32's

> Do you have idea why are values from 2bytes regiters read wrong?

If it's a register showing as int16 us REGISTER_S16

**Comment by jiriknez** [comment]
![Solax_screen](https://github.com/user-attachments/assets/2446b622-2de4-457d-825b-aa0fd81f618d)

If I use int16 insted of int32, value is 0 as you can see on Active Power (register 183). Second line is read as REGISTER_S32 and value is nonsense (register 185). Values (Voltage and Current) with 1 bytes length are ok.

**Comment by wills106** [comment]
I meant the PV ones that are showing as int16 in the docs, but are uint16 in the Integration.

I have just pushed out [2024.08.3](https://github.com/wills106/homeassistant-solax-modbus/releases/tag/2024.08.3) with corrections to match the docs.

Just so you know I have renamed "Today Energy" to "Today's Energy" to match the other Inverters.

**Comment by jiriknez** [comment]
I don't know what's wrong now, because i see every sencor as Unknown, but it doesn't report any error. Everything seems to be ok in log, values are read.  

Please correct:
register 0x28 is used 2x
name and key reactive_power is also 2x

**Comment by wills106** [comment]
I take it you meant 0x28F?

Just released [2024.08.4](https://github.com/wills106/homeassistant-solax-modbus/releases/tag/2024.08.4)

Had to change names on the active / reactive control.

**Comment by wills106** [comment]
> no local data file found after 5 tries

That can be ignored. It's only used on the Gen4 & 5 SolaX Hybrids.



> And pls add "register_type = REG_INPUT" to Model Type and Software Version.

Have you corrected that locally? If so what do they return?

**Comment by jiriknez** [comment]
I corrected it locally, but I still see Unknown and I'm not able to figure out, why. 
![Solax_unknown](https://github.com/user-attachments/assets/75e24d2f-a0b4-4757-b6bc-ebaa46fa488e)

**Comment by wills106** [comment]
If you go to the detail error log does it say it's failing on one of the registers?

If it doesn't try deleting the whole block for "Model Type", if it still doesn't work try deleting the "Software Version" as well.

"Model Type" is Red in the Modbus docs, so might not be there on all versions?

**Comment by jiriknez** [comment]
I found it. Register 0x18F was set as holding. I'm sending corrected plugin. 
[plugin_solax_mega_forth(1).zip](https://github.com/user-attachments/files/16487613/plugin_solax_mega_forth.1.zip)

But, back to the start. Values from 2bytes registers are still wrong. Do you have any idea what to do with it.
![Solax_sensors](https://github.com/user-attachments/assets/970e5990-7573-4eb2-91bd-6c04fb089e22)

**Comment by wills106** [comment]
Try changing `LITTLE` to` BIG`

https://github.com/wills106/homeassistant-solax-modbus/blob/2389c7cb058b5eae527e0a4b2cc321bca7f5b490/custom_components/solax_modbus/plugin_solax_mega_forth.py#L674

**Comment by jiriknez** [comment]
Yes, that's it. Thanks a lot. Now is everything works perfectly.

**Comment by wills106** [comment]
Does "Model Type" & "Software Version" return anything useful?

If they do then I can add them to the Device Info panel.

Is your Inverter a 40kW version with 4 MPPT's or a larger Inverter?

**Comment by jiriknez** [comment]
Yes, it does.
![Solax_diag](https://github.com/user-attachments/assets/c4eb1be5-3c1f-4618-88fd-95983aa681ae)

This tested inverter is X3-40 with 4 MPPT's. But we have also installed a lot of X3-50, X3-60 and X3-100 (110). I will integrate all of them to HA. If I will add sensors in plugin, I'll send it to you.

**Comment by wills106** [comment]
I'm in the progress of adding in the other MPPT's

Just need to find the first 6 or so digits of each power type. So I can show the extra MPPT's only on the larger systems.

**Comment by jiriknez** [comment]
X3G060 - X3-60 - 6 MPPT
X3G050 - X3-50 - 5 MPPT

X3-FORTH-80-110K - 9 MPPT, but now I don't know what SN starts with.

**Comment by wills106** [comment]
I thought the 50 & 60 would be like that.

75 onwards could start using X3G075 or X3G75 hard to tell till we get reports of others.

**Comment by jiriknez** [comment]
I think that Solax have only X3-MGA-(40-60) and X3-FTH-(80-125), none 75kW.

**Comment by wills106** [comment]
It's mentioned in the Datasheet. But maybe no one sells them as so close to the 80kw?

**Comment by wills106** [comment]
Don't suppose you have access to the Modbus docs for the new X1-SMART-G2?

https://github.com/wills106/homeassistant-solax-modbus/discussions/520#discussioncomment-10110669

**Comment by jiriknez** [comment]
I can ask for it, it's not problem. I can send it to you in two days.

**Comment by wills106** [comment]
[2024.08.5](https://github.com/wills106/homeassistant-solax-modbus/releases/tag/2024.08.5) contains the extra MPPT's.
It will always return the set amount for Inverter type even if some MPPT's are not connected.

I don't know if input register 0x280 returns a dynamic count, or if that's always fixed as well.

**Comment by wills106** [comment]
Just realised there is an error in the setup. On a 60kW it will only display MPPT 1-4 & 6.
I need to correct it so it displays 5 as well.

Same on a 150kW. Will display MPPT 1-4 & 10-12

**Comment by wills106** [comment]
Corrected in [2024.08.6](https://github.com/wills106/homeassistant-solax-modbus/releases/tag/2024.08.6)

**Comment by jiriknez** [comment]
> I don't know if input register 0x280 returns a dynamic count, or if that's always fixed as well. 

I don't know either. After I will connect next inverters, I'll let you know.

After update, it works good.

**Comment by wills106** [comment]
How complete do you think the support is for X3 Mega Inverters? Can this be closed down now?

**Comment by jiriknez** [comment]
Yes, it can be closed.

**Comment by MultissCZ** [comment]
Hello,

I would like to ask about the status of the functionality for the Solax X3 Forth 120kW.

I need to not only read data from it but also write data. Reading works, but writing does not. Is there any issue with this?

Thank you for your response.

**Comment by kyselyna** [comment]
hello 

here is the latest modbus protocol with error codes

[Solax Power X3-FORTH Modbus-RTU comms protocol（2025.4.21）.pdf](https://github.com/user-attachments/files/22369131/Solax.Power.X3-FORTH.Modbus-RTU.comms.protocol.2025.4.21.pdf)

**Comment by wills106** [comment]
Thanks @kyselyna have you found any other new registers that might be interest to implement?

**Comment by kyselyna** [comment]
no problem I think it would be interesting to have these registers implemented

<img width="767" height="54" alt="Image" src="https://github.com/user-attachments/assets/d507dbce-243e-46f5-89a1-80edf1170f54" />

<img width="799" height="108" alt="Image" src="https://github.com/user-attachments/assets/3519e537-1611-4693-aff2-ec7be0cf5781" />

It would also be great if it were possible to set a limit on the inverter's export power when an electricity meter is connected and see the actual consumption of the object, but I couldn't find these registers.
